beacon_node/eth1/src/deposit_cache.rs

```diff
@@ -21,21 +21,83 @@ pub enum Error {
     /// A log with the given index is already present in the cache and it does not match the one
     /// provided.
     DuplicateDistinctLog(u64),
+    /// Attempted to insert log with given index after the log had been finalized
+    FinalizedLogInsert {
+        log_index: u64,
+        finalized_index: u64,
+    },
     /// The deposit count must always be large enough to account for the requested deposit range.
     ///
     /// E.g., you cannot request deposit 10 when the deposit count is 9.
     DepositCountInvalid { deposit_count: u64, range_end: u64 },
+    /// You can't request deposits on or before the finalized deposit
+    DepositRangeInvalid {
+        range_start: u64,
+        finalized_count: u64,
+    },
+    /// You can't finalize what's already been finalized and the cache must have the logs
+    /// that you wish to finalize
+    InvalidFinalizeIndex {
+        requested_count: u64,
+        currently_finalized: u64,
+        deposit_count: u64,
+    },
     /// Error with the merkle tree for deposits.
     DepositTree(merkle_proof::MerkleTreeError),
     /// An unexpected condition was encountered.
     Internal(String),
 }
 
+#[derive(Encode, Decode, Clone)]
+pub struct SszLegacyDepositCache {
+    logs: Vec<DepositLog>,
+    leaves: Vec<Hash256>,
+    deposit_contract_deploy_block: u64,
+    deposit_roots: Vec<Hash256>,
+}
+
+// TODO: delete this after the new SszDepositCache has been release for a while
+impl SszLegacyDepositCache {
+    pub fn to_deposit_cache(&self) -> Result<DepositCache, String> {
+        let deposit_tree =
+            DepositDataTree::create(&self.leaves, self.leaves.len(), DEPOSIT_TREE_DEPTH);
+        // Check for invalid SszDepositCache conditions
+        if self.leaves.len() != self.logs.len() {
+            return Err("Invalid SszDepositCache: logs and leaves should have equal length".into());
+        }
+        // `deposit_roots` also includes the zero root
+        if self.leaves.len() + 1 != self.deposit_roots.len() {
+            return Err(
+                "Invalid SszDepositCache: deposit_roots length must be only one more than leaves"
+                    .into(),
+            );
+        }
+        let finalized_block_height = if self.deposit_contract_deploy_block == 0 {
+            0
+        } else {
+            self.deposit_contract_deploy_block - 1
+        };
+
```

Comment:
This can be written more succinctly as `self.deposit_contract_deploy_block.saturating_sub(1)`.

Reply:
done!